### PR TITLE
[SourceKit] Resolve info request

### PR DESCRIPTION
**Pull request by skrtks**
This adds a new request to SourceKit that provides resolve information for the references in a file. It returns the path to the document, the offset, and the length of the declaration if this information is available. For the declarations that are not in the 
project sources, it returns the USR and the module name. `tools/SourceKit/docs/Protocol.md` contains a detailed description of the request.

@rintaro do you mind taking a look at this?

### CODE REVIEW

**Comment by skrtks** [comment]
Added USR to the response for declarations that aren't in generated interfaces and updated tests accordingly.

File: test/SourceKit/ResolveInfo/Inputs/other.swift [hunk 1]
```diff
@@ -0,0 +1,6 @@
+extension C {
```
Comment (bnbarham) [inline test/SourceKit/ResolveInfo/Inputs/other.swift:1]: Could you change the tests to have 2-space indenting? We should really automatically format these but we don't have that infrastructure at the moment 😅

**Review by bnbarham — COMMENTED** [review]
Few high level thoughts:
 - We're moving towards using the constraint solver for the current cursor info. That means that it will return different results to those from a `SourceEntityWalker` walk (to allow giving back results when they're ambiguous/there's errors) + doesn't require a full typecheck. Returning *all* results on every change (or after a delay more likely) also seems like a lot to me - especially when the majority won't actually be used.
 - Cursor info returns more information, eg. whether a reference is "dynamic" (could resolve to a different override/is an existential) and the USR of the static type. Reasonable to leave that out if you don't need it though.
 - Initializers aren't the only "secondary decl" case. Consider eg. shorthand if bindings (`if let foo {`) and more generally, all the current cases in cursor info. I don't think we have any that return multiple secondaries at the moment, but IMO it would be better to just allow any number rather than primary + secondary. The cursor info API is the way it is so that clients didn't have to be updated when the secondary decls were added.

Another thought, this is basically doing what the index does. Have you considered just running the indexer and processing that output? There's an option to index locals as well (`-index-include-locals`), so I believe that would have everything you need.

**Comment by skrtks** [comment]
Thanks for the feedback. I didn't know about the `-index-include-locals`. If I update our internal strategy a bit, I think it might be a better fit than the request I was working on. I will close this PR for now.

**Pull request closed without merging**